core/blockchain.go

```diff
@@ -298,15 +305,17 @@ func NewBlockChain(db ethdb.Database, cacheConfig *CacheConfig, chainConfig *par
 			}
 		}
 		if needRewind {
-			var hashes []common.Hash
-			previous := bc.CurrentHeader().Number.Uint64()
-			for i := low + 1; i <= bc.CurrentHeader().Number.Uint64(); i++ {
-				hashes = append(hashes, rawdb.ReadCanonicalHash(bc.db, i))
+			log.Error("Truncating ancient chain", "from", bc.CurrentHeader().Number.Uint64(), "to", low)
+			if err := bc.SetHead(low); err != nil {
```

Comment:
Here exists an issue. 

If we are in the fast sync and then crash:

- Ancient blocks are written
- Fast block marker is not updated
- We need to truncate the extra ancient blocks
- We call `SetHead(low)` where the `HeadBlock` is genesis
- We enter the here

```go
// Rewind the chain to the requested head and keep going backwards until a
// block with a state is found or fast sync pivot is passed
log.Warn("Rewinding blockchain", "target", head)
bc.hc.SetHead(head, updateFn, delFn)
``` 

During the rewinding, the `newHead` is set to `genesis` which just delete the entire chain.